pyomo/util/slices.py

```diff
@@ -0,0 +1,275 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import six
+
+from pyomo.core.base.indexed_component import normalize_index
+from pyomo.core.base.indexed_component_slice import IndexedComponent_slice
+from pyomo.core.base.global_set import UnindexedComponent_set
+from pyomo.common.collections import ComponentSet, ComponentMap
+
+def _to_iterable(source):
+    iterable_scalars = six.string_types + (six.binary_type, six.text_type)
+    if isinstance(source, iterable_scalars):
+        yield source
+    elif hasattr(source, '__iter__'):
+        for obj in source:
+            yield obj
```

Comment:
For efficiency, it might make sense to switch this:
```python
if hasattr(source, '__iter__'):
    if isinstance(source, iterable_scalars):
        yield source
    for obj in source:
        yield obj
```
This reduces the umber of `isinstance` calls that are expected to return False (which can be slow).

Reply:
Easy enough, and makes sense given the high volume of non-iterable scalars we expect.